From 18b8c72fd9690d0ec774cfa35b825c8accba4052 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:29:06 -0400
Subject: [alsa] ice1712: Add suspend support for M-Audio ICE1712-based cards

Message-id: <140972823573-249-patchset-jkysela@redhat.com>
Patchwork-id: 90949
O-Subject: [RHEL 7.1 PATCH 248/442] ALSA: ice1712: Add suspend support for M-Audio ICE1712-based cards
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

Add suspend support for M-Audio cards based on ICE1712 chip.
Tested with M-Audio Audiophile 24/96. S/PDIF will probably not work.

Signed-off-by: Ondrej Zary <linux@rainbow-software.org>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 8c1d843460f42417d6b9553147a1a04ca1470602)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/ice1712/delta.c b/sound/pci/ice1712/delta.c
index 98e5e95..ed2144e 100644
--- a/sound/pci/ice1712/delta.c
+++ b/sound/pci/ice1712/delta.c
@@ -576,6 +576,30 @@ static struct snd_ak4xxx_private akm_vx442_priv = {
  .mask_flags = 0,
 };
 
+#ifdef CONFIG_PM_SLEEP
+static int snd_ice1712_delta_resume(struct snd_ice1712 *ice)
+{
+ unsigned char akm_backup[AK4XXX_IMAGE_SIZE];
+ /* init codec and restore registers */
+ if (ice->akm_codecs) {
+  memcpy(akm_backup, ice->akm->images, sizeof(akm_backup));
+  snd_akm4xxx_init(ice->akm);
+  memcpy(ice->akm->images, akm_backup, sizeof(akm_backup));
+  snd_akm4xxx_reset(ice->akm, 0);
+ }
+
+ return 0;
+}
+
+static int snd_ice1712_delta_suspend(struct snd_ice1712 *ice)
+{
+ if (ice->akm_codecs) /* reset & mute codec */
+  snd_akm4xxx_reset(ice->akm, 1);
+
+ return 0;
+}
+#endif
+
 static int snd_ice1712_delta_init(struct snd_ice1712 *ice)
 {
  int err;
@@ -622,7 +646,11 @@ static int snd_ice1712_delta_init(struct snd_ice1712 *ice)
   ice->num_total_adcs = 4;
   break;
  }
-
+#ifdef CONFIG_PM_SLEEP
+ ice->pm_resume = snd_ice1712_delta_resume;
+ ice->pm_suspend = snd_ice1712_delta_suspend;
+ ice->pm_suspend_enabled = 1;
+#endif
  /* initialize the SPI clock to high */
  tmp = snd_ice1712_read(ice, ICE1712_IREG_GPIO_DATA);
  tmp |= ICE1712_DELTA_AP_CCLK;
-- 
1.7.1